### PR TITLE
Make De-duplication Multi-threaded and Happen Only During Pre-processing

### LINKED ISSUE #2719: Save De-duplicated Set During Pre-processing
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Currently the de-duplication is being run *after* the set has been saved during the pre-processing phase, and gets run again when starting the training phase (as in doing `-m axolotl.cli.preprocess` then doing `-m axolotl.cli.train`).

For small sets it might not matter much, but for larger sets (which would make more sense to be de-duplicating) this takes a *long* time. That means that if you had a large dataset that took like 15 minutes to de-duplicate you'd be spending 30 minutes on just de-duplicating, before you get to take a single training step, plus another 15 minutes any time you wanted to restart training for whatever reason, like OOM, hyper-parameter tuning, etc.

```
Example time to de-duplicate one of my configs: ~14 minutes.

Saving the dataset (23/23 shards): 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 162329/162329 [01:35<00:00, 1694.97 examples/s]
[2025-05-23 22:41:55,123] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:134] [PID:122699] [RANK:0] Train dataset is None. Skipping deduplication.
[2025-05-23 22:41:55,123] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:147] [PID:122699] [RANK:0] Eval dataset is None. Skipping deduplication.
[2025-05-23 22:41:55,124] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:150] [PID:122699] [RANK:0] Starting deduplication for combined dataset. Original size: 162329
[2025-05-23 22:56:28,138] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:154] [PID:122699] [RANK:0] Deduplication complete for combined dataset. New size: 159750
```

If this de-duplication step was instead done *only* during the pre-processing stage and then *saved*, you would only need to do it once and would also be able to push the de-duplicated set to huggingface.

### ✔️ Solution

Move the de-duplication to *before* saving in the pre-processing phase. Also prevent it from being run before the training phase.

### ❓ Alternatives

Not exactly an alternative, but an addition; making the de-duplication code multi-threaded. Currently it only seems to hit 1 or 2 threads. So that's probably where a lot of the slowness is coming from.

### 📝 Additional Context

_No response_

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by xzuyn**
Uses only the SHA256 hash. Collision is too unlikely to be worth checking for.

262K samples de-duped in ~4 minutes. Previously it took ~15 minutes to de-dupe 162K samples. So roughly 6x faster, or even faster if you count it de-duping multiple times before.

```
[2025-06-01 06:26:34,020] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:132] [PID:74277] [RANK:0] Train dataset is None. Skipping deduplication.
[2025-06-01 06:26:34,020] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:145] [PID:74277] [RANK:0] Eval dataset is None. Skipping deduplication.
[2025-06-01 06:26:34,020] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:148] [PID:74277] [RANK:0] Starting deduplication for combined dataset. Original size: 262437
Map (num_proc=12): 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 262437/262437 [03:45<00:00, 1163.76 examples/s]
[2025-06-01 06:30:20,862] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:152] [PID:74277] [RANK:0] Deduplication complete for combined dataset. New size: 258523
```

Resolves #2719

## Summary by CodeRabbit

- **Refactor**
	- Improved dataset deduplication with faster, parallel processing using hash-based methods.
	- Deduplication now occurs once immediately after dataset merging, reducing redundant steps.
- **Chores**
	- Simplified dataset preparation workflow for better consistency and efficiency.

### CODE REVIEW

File: src/axolotl/utils/data/utils.py [hunk 1]
```diff
@@ -69,39 +70,31 @@ def sha256(to_hash: str, encoding: str = "utf-8") -> str:
...
+    if other_dataset is not None:
+        other_hashed = other_dataset.map(compute_row_hash, num_proc=num_proc)
+        other_hashes = set(other_hashed["row_hash"])
+
+    for idx, row_hash in enumerate(hashes):
+        if row_hash in seen_hashes or row_hash in other_hashes:
+            continue
+        seen_hashes.add(row_hash)
+        unique_indices.add(idx)
+
```
Comment (winglian) [inline src/axolotl/utils/data/utils.py:111]: It might be worth dropping the added `row_hash` column here now that it's done.. The previous implementation did not modify the dataset.

Comment (xzuyn) [reply src/axolotl/utils/data/utils.py:111]: this shouldn't be modifying it either. the only change that happens to `dataset` is at the end when it returns with `.select`. the `row_hash` stuff comes from a copy in `hashed` and `other_hashed` (which I should probably make delete after grabbing the hashes)

File: src/axolotl/utils/data/utils.py [hunk 2]
```diff
@@ -69,39 +70,39 @@ def sha256(to_hash: str, encoding: str = "utf-8") -> str:
...
-                    if original_idx < len(other_dataset) and str(
-                        other_dataset[original_idx]
-                    ) == str(row):
-                        is_duplicate = True
-                        break
-            if not is_duplicate:
-                seen_hashes[row_hash].append(idx)
-                unique_indices.append(idx)
-                continue
+    hashes, other_hashes, seen_hashes, unique_indices = [], set(), set(), set()
```
[the remark this answers is not in the document]
Comment (xzuyn) [reply src/axolotl/utils/data/utils.py]: not correct. it expects an `Iterable`, which a set is. the code runs correctly.

**Comment by winglian** [comment]
@djsaunde does #2707 overlap or supersede this?

**Comment by djsaunde** [comment]
> @djsaunde does #2707 overlap or supersede this?

Should go in after.
